### PR TITLE
Coupled value function IC

### LINKED ISSUE #16405: Coupled variable function IC
## Reason
I need to set ICs for variables based on the values of other variables. An easy way would be an IC that allows me to couple in a set of (up to 4 variable) the values of which are then substituted into a moose function (into x,y,z, and t) and the function value is taken as the IC. This woudl allow me to use PiecewiseMultiLinear data to interpolate tabulated datasets.

## Design
Add IC that can couple in variables and use them to evaluate a function to determine the IC value.

## Impact
Added capability

### PR DESCRIPTION
**Pull request by dschwen**
Initial condition that can the values of up to four coupled variables processed through a MOOSE function to set the value of a variable.

Closes #16405

### CODE REVIEW

File: modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# CoupledValueFunctionIC
+
+!syntax description /ICs/CoupledValueFunctionIC
+
+CoupledValueFunctionIC initializes the value of a variable with the value of a
+coupled  [MOOSE Function](functions.md) which is evaluated with a set of up to
```
Comment (loganharbour) [inline modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md:6]:
```suggestion
coupled [MOOSE Function](functions.md), which is evaluated with a set of up to
```

File: modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+four coupled  variable values `v` as its input parameters. The coupled variable
```
Comment (loganharbour) [inline modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md:7]:
```suggestion
four coupled variable values `v` as its input parameters. The coupled variable
```

File: modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+values are sunstituted for the  `x`,`y`,`y`, and `t` function variables in that
```
Comment (loganharbour) [inline modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md:8]:
```suggestion
values are substituted for the `x`,`y`,`y`, and `t` function variables in that
```

File: modules/phase_field/src/ics/CoupledValueFunctionIC.C [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+InputParameters
+CoupledValueFunctionIC::validParams()
+{
+  InputParameters params = InitialCondition::validParams();
+  params.addClassDescription("Initialize the variable from a lookup function");
+  params.addRequiredParam<FunctionName>("function",
+                                        "Coupled function to evaluate with values from v");
+  params.addCoupledVar("v",
+                       "List of up to four coupled variables that are substituted for x,y,z, and t "
+                       "in the coupled function");
```
Comment (loganharbour) [inline modules/phase_field/src/ics/CoupledValueFunctionIC.C:24]: I'm a little concerned about putting these in a specific order. 

Is there no case where one would need to provide only x, y, and t? Would seem silly to have to provide a dummy for z.

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: modules/phase_field/src/ics/CoupledValueFunctionIC.C (hunk 4 above)
Comment (dschwen) [reply modules/phase_field/src/ics/CoupledValueFunctionIC.C:24]: Hm, the application I have in mind is creating a special mapping function for a n variable problem. For a three variable problem you'd just always use x,y,z. Making that configurable seems like unnecessary complexity.

Comment (loganharbour) [reply modules/phase_field/src/ics/CoupledValueFunctionIC.C:24]: Alright. That's fine with me, as long as it was kept in mind with the design. 

I'm good to approve once the doc changes are added and docs pass.

File: modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md (hunk 3 above)
Comment (GiudGiud) [inline modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md:8]:
```suggestion
values are sunstituted for the  `x`,`y`,`z`, and `t` function variables in that
```

File: modules/phase_field/include/ics/SpecifiedSmoothCircleIC.h [hunk 5]
```diff
@@ -11,14 +11,11 @@
...
-// Forward Declarations
-
 /**
- * SpecifiedsmoothCircleIC creates multiple SmoothCircles (number = size of x_positions) that are
- *positioned in the
- * set locations with the set radii.  This is adapted from PolySpecifiedSmoothCircleIC from HYRAX by
- *A.M. Jokisaari
- **/
+ * SpecifiedsmoothCircleIC creates multiple SmoothCircles (number = size of x_positions)
+ * that arepositioned in the set locations with the set radii.
```
Comment (GiudGiud) [inline modules/phase_field/include/ics/SpecifiedSmoothCircleIC.h:16]:
```suggestion
 * that are positioned in the set locations with the set radii.
```

File: modules/phase_field/test/tests/misc/coupled_value_function_ic.i [hunk 6]
```diff
@@ -0,0 +1,79 @@
+[Mesh]
+  [gen]
+    type = GeneratedMeshGenerator
+    dim = 2
+    nx = 10
+    ny = 10
+  []
+[]
+
+# Here we sum up the inverses of teh ICs above. This should add up to 2.0 everywhere
```
Comment (GiudGiud) [inline modules/phase_field/test/tests/misc/coupled_value_function_ic.i:10]:
```suggestion
# Here we sum up the inverses of the ICs above. This should add up to 2.0 everywhere
```

File: modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md (hunk 3 above)
Comment (loganharbour) [reply modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md]: Lol did you even look at my review?

Comment (GiudGiud) [reply modules/phase_field/doc/content/source/ics/CoupledValueFunctionIC.md]: I think so?
I didnt report the typo you reported

**Comment by dschwen** [comment]
Gentlemen, I believe I addressed your comments.

**Review by loganharbour — APPROVED** [review]
Sir, it would appear that you indeed have.

**Pull request merged**